exporter/exporter.go

```diff
@@ -0,0 +1,431 @@
+					}
+
+					values, err := strconv.ParseFloat(seps[1], 64)
+					if err != nil {
+						continue
+					}
+
+					if metric, ok := exporter.metadMap[seps[0]]; ok {
+						ch <- prometheus.MustNewConstMetric(metric.desc, prometheus.GaugeValue, values,
+							item.InstanceName,
+							"nebula",
+							"metad",
+						)
+					}
+				}
+
+			} else if item.NebulaType == "storaged" {
+				log.Infof("Begin Collect Storaged %s Metrics ", item.InstanceName)
+				for _, singleMetric := range metrics {
+					log.Info(singleMetric)
+					seps := strings.Split(singleMetric, "=")
+					if len(seps) != 2 {
+						continue
+					}
+
+					values, err := strconv.ParseFloat(seps[1], 64)
+					if err != nil {
+						continue
+					}
+
+					if metric, ok := exporter.storagedMap[seps[0]]; ok {
+						ch <- prometheus.MustNewConstMetric(metric.desc, prometheus.GaugeValue, values,
+							item.InstanceName,
+							"nebula",
+							"storaged",
+						)
+					}
+				}
+			} else if item.NebulaType == "graphd" {
+				log.Infof("Begin Collect Graphd %s Metrics ", item.InstanceName)
+				for _, singleMetric := range metrics {
+					log.Info(singleMetric)
+					seps := strings.Split(singleMetric, "=")
+					if len(seps) != 2 {
+						continue
+					}
+
+					values, err := strconv.ParseFloat(seps[1], 64)
+					if err != nil {
+						continue
+					}
+
+					if metric, ok := exporter.graphdMap[seps[0]]; ok {
+						ch <- prometheus.MustNewConstMetric(metric.desc, prometheus.GaugeValue, values,
+							item.InstanceName,
+							"nebula",
+							"graphd",
+						)
+					}
+				}
+			}
+			wg.Done()
+		}()
+
+		wg.Wait()
+		log.Info("Collect Nebula Metrics Done")
+	}
+}
+
+func (exporter *NebulaExporter) CollectFromKuberneres(ch chan<- prometheus.Metric) {
+	log.Info("Begin Collect Nebula Metrics")
+	podLists, err := exporter.Client.CoreV1().Pods(exporter.Namespace).List(metav1.ListOptions{})
+
+	if err != nil {
+		return
+	}
+
+	var wg sync.WaitGroup
+
+	for _, pod := range podLists.Items {
+		podIpAddress := pod.Status.PodIP
+		podHttpPort := int32(0)
+		for _, port := range pod.Spec.Containers[0].Ports {
+			if port.Name == "http" {
+				podHttpPort = port.ContainerPort
+			}
+		}
+
+		if podHttpPort == 0 {
+			continue
+		}
+
+		wg.Add(1)
+
+		go func() {
+			metrics, err := getNebulaMetrics(podIpAddress, podHttpPort)
+
+			if err != nil {
+				wg.Done()
+				return
+			}
+			if strings.Contains(pod.Name, "metad") {
+				log.Infof("Begin Collect Metad %s Metrics ", pod.Name)
+				for _, singleMetric := range metrics {
+					log.Info(singleMetric)
+					seps := strings.Split(singleMetric, "=")
+					if len(seps) != 2 {
+						continue
+					}
+
+					values, err := strconv.ParseFloat(seps[1], 64)
+					if err != nil {
+						continue
+					}
+
+					if metric, ok := exporter.metadMap[seps[0]]; ok {
+						ch <- prometheus.MustNewConstMetric(metric.desc, prometheus.GaugeValue, values,
+							pod.Name,
+							pod.Namespace,
+							"metad",
+						)
+					}
+				}
```

Comment:
Why not to extract some common functions for collecting `graph`/`meta`/`storage` metrics from static and k8s server ?

Reply:
this is done in #2,  extrace a function to collect metrics